locallib.php

```diff
@@ -328,3 +332,50 @@ function sendfailurenotification( $admininfo) {
     return ('Email sent to ' . $admininfo->email);
 }
 
+/**
+ * Enable communication with Arlo and reset the redirect counter of the plugin.
+ */
+function enrol_arlo_enablecommunication() {
+    $plugin = api::get_enrolment_plugin();
+    $pluginconfig = $plugin->get_plugin_config();
+    $pluginconfig->set('enablecommunication', 1);
+    $pluginconfig->set('redirectcount', 0);
+}	
+
+/**
+ * Reset the redirect counter for all registrations that have reached the maximum number of retries.
+ */
+function enrol_arlo_reset_redirects() {
+    global $DB;
+    $plugin = api::get_enrolment_plugin();
+    $pluginconfig = $plugin->get_plugin_config();
+    $pluginconfig->set('redirectcount', 0);
+    $maxretries = get_config('enrol_arlo', 'retriesperrecord');
+    $sql = "SELECT * FROM {enrol_arlo_registration} 
+                    WHERE redirectcounter >= :maxretries
+                          AND updatesource = 1";
+    $registrations = $DB->get_records_sql($sql, ['maxretries' => $maxretries]);
+    foreach ($registrations as $registration) {
+        $registration->redirectcounter = 0;
+        $DB->update_record('enrol_arlo_registration', $registration);
+    }
+}
+
+/**
+ * Update all registrations in a course that have not been graded or have no outcome.
+ *
+ * @param $courseid
+ */
+function enrol_arlo_update_all_course_registrations($courseid) {
+    global $DB;
+    $enrolid = $DB->get_field('enrol', 'id', ['enrol' => 'arlo', 'courseid' => $courseid]);
```

Comment:
I would transform this in one query. If you can have one query instead of two...

```sql
SELECT ear.*
FROM {enrol_arlo_registration} ear
JOIN {enrol} e ON e.enrol = 'arlo' AND e.courseid = :courseid
WHERE ear.enrolid = e.id
  AND ear.grade IS NULL
  AND ear.outcome IS NULL
```

Reply:
This makes a lot of sense, thanks Darko